File: 프로그래머스/lv3/43163. 단어 변환/단어 변환.py
def isDifferenceOne(a, b):
    diffCnt = 0

    for i in range(len(a)):
        if a[i] != b[i]:
            diffCnt += 1

    return diffCnt == 1


def dfs(words, target, word, count, route):
        global answer
        global chk
        if word == target:
            answer.append(count)
            return
        
        for i in range(len(words)):
            if isDifferenceOne(words[i], word) and chk[i] != 1:
                chk[i] = 1
                dfs(words, target, words[i], count + 1, route + [words[i]])
                chk[i] = 0
                
    
    
def solution(begin, target, words):
    global answer
    global chk
    answer = []
    chk = [0] * len(words)
    
    if target not in words:
        return 0
    
    for idx, w in enumerate(words):
        count = 0
        changeableWord = begin

        if isDifferenceOne(w, changeableWord) and chk[idx] != 1:    
            changeableWord = w
            chk[idx] = 1    
                    
            if changeableWord == target:
                return count + 1
            
            dfs(words, target, changeableWord, count + 1, [begin, w])
            chk[idx] = 0
            count -= 1

    if answer:
        return min(answer)
    else:
        return 0

File: 프로그래머스/3/43163. 단어 변환/단어 변환.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;
int answer = 0;

bool is_difference_once(string s1, string s2) {
    int count = 0;

    for (int i = 0; i < s1.size(); i++) {
        if (s1[i] != s2[i]) {
            count++;
        }
    }

    return count == 1;
}

void dfs(string current_string, string target, vector<string> words, vector<bool> & chk, int count) {
    if (current_string == target) {
        answer = count;
        return;
    }

    for (int i = 0; i < words.size(); i++) {
        if (!chk[i] && is_difference_once(current_string, words[i])) {
            chk[i] = true;
            dfs(words[i], target, words, chk, count + 1);
            chk[i] = false;
        }
    }
}

int solution(string begin, string target, vector<string> words) {
    if (find(words.begin(), words.end(), target) == words.end()) {
        return 0;
    }

    vector<bool> chk(words.size(), false);

    dfs(begin, target, words, chk, 0);
    return answer;

}